### PR TITLE
Realize unified Azure Spring token credentials

### PR DESCRIPTION
**Pull request by saragluna**
The SDK-supported credential table: https://github.com/Azure/azure-sdk-for-java/wiki/Spring:-Unify-Azure-Credentials-construction-in-Azure-Spring. 

In a Spring Boot application, we will provide a `SpringEnvironmentCredential`, the Spring version of `EnvironmentCredential`, which will read credential properties via Spring properties and default environment variables defined in Azure SDK. We'll also delegate all TokenCredential defined in `azure-identity` for we can't extend the `CredentialBuilderBase`. In the `AzureContextCredentialAutoConfiguration`, we will define several `SpringCredentialBuilderBase` implementations, order by their precedence, and configure the identity options in a bean factory post-processor, then configure all the credentials in a `ChainedTokenCredentialBuilder` at last.

When an Azure Spring starter user wants to customize their own TokenCredential, he or she can provide a `TokenCredential` bean directly or inject a `SpringCredentialBuilderBase` implementation with higher precedence. An `IdentityOptions` bean could be defined to customize the identity options.

### CODE REVIEW

File: sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/DefaultSpringCredentialBuilder.java [hunk 1]
```diff
@@ -15,54 +26,52 @@
...
 
-    /**
-     * Defines the AZURE_CREDENTIAL_PREFIX.
-     */
-    static final String AZURE_CREDENTIAL_PREFIX = "azure.credential.";
+    private String managedIdentityClientId;
+    private String tenantId;
 
-    private String alternativePrefix;
+    public DefaultSpringCredentialBuilder(Environment environment) {
```
Comment (chenrujun) [inline sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/DefaultSpringCredentialBuilder.java:32]: Builders like this didn't bring much benefits, how about deleting these builders and use constructor instead?

File: sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/DefaultSpringCredentialBuilder.java [hunk 2]
```diff
@@ -15,54 +26,52 @@
...
-     * managed identity credential.
+     * Build a {@link DefaultSpringCredential} with below list of credentials:
+     * <ol>
+     * <li>{@link SpringEnvironmentCredential}</li>
+     * <li>{@link ManagedIdentityCredential}</li>
+     * <li>{@link SharedTokenCacheCredential}</li>
+     * <li>{@link IntelliJCredential}</li>
+     * <li>{@link VisualStudioCodeCredential}</li>
+     * <li>{@link AzureCliCredential}</li>
+     * <li>{@link AzurePowerShellCredential}</li>
```
Comment (chenrujun) [inline sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/DefaultSpringCredentialBuilder.java:58]: After using this builder, we should keep the java doc twice, it's error-prone: forget to update one place when update the java doc.

File: sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/PrefixedSpringEnvironmentCredential.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                        "variables are not fully configured.");
+                LOGGER.error(exception.getMessage(), exception);
+                return exception;
+            });
+        } else {
+            return tokenCredential.getToken(request);
+        }
+    }
+
+    private TokenCredential populateTokenCredential(String prefix) {
```
Comment (chenrujun) [inline sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/PrefixedSpringEnvironmentCredential.java:56]: Some duplicated logic in `PrefixedSpringEnvironmentCredential#populateTokenCredential` and `SpringEnvironmentCredential#populateTokenCredential`, could you please reuse them?

File: sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/PrefixedSpringEnvironmentCredential.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        String tenantId = azureEnvironment.getPropertyValue(prefix + TENANT_ID);
+        String clientId = azureEnvironment.getPropertyValue(prefix + CLIENT_ID);
+        String clientSecret = azureEnvironment.getPropertyValue(prefix + CLIENT_SECRET);
+
+        if (tenantId != null && clientId != null && clientSecret != null) {
```
Comment (chenrujun) [inline sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/PrefixedSpringEnvironmentCredential.java:61]: How about adding logs here to let customer know which kind of credential is populated?

File: sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/PrefixedSpringEnvironmentCredential.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            return new ClientSecretCredentialBuilder().tenantId(tenantId)
+                                                      .clientId(clientId)
+                                                      .clientSecret(clientSecret)
+                                                      .build();
+        }
+
+        String certPath = azureEnvironment.getPropertyValue(prefix + CLIENT_CERTIFICATE_PATH);
+
+        if (tenantId != null && clientId != null && certPath != null) {
```
Comment (chenrujun) [inline sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/PrefixedSpringEnvironmentCredential.java:70]: Same here.

File: sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/PrefixedSpringEnvironmentCredential.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            return new ClientCertificateCredentialBuilder().tenantId(tenantId)
+                                                           .clientId(clientId)
+                                                           .pemCertificate(certPath)
+                                                           .build();
+        }
+
+        String username = azureEnvironment.getPropertyValue(prefix + USERNAME);
+        String password = azureEnvironment.getPropertyValue(prefix + PASSWORD);
+
+        if (clientId != null && username != null && password != null) {// tenant-id is not required
```
Comment (chenrujun) [inline sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/PrefixedSpringEnvironmentCredential.java:80]: Here.

File: sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/PrefixedSpringEnvironmentCredentialBuilder.java [hunk 7]
```diff
@@ -3,16 +3,22 @@
...
-import org.springframework.util.StringUtils;
+import org.springframework.core.env.Environment;
+import org.springframework.util.Assert;
 
 /**
  * Spring token credential built from the prefixed properties.
  */
-public class PrefixedSpringCredentialBuilder extends SpringCredentialBuilderBase<PrefixedSpringCredentialBuilder> {
+public class PrefixedSpringEnvironmentCredentialBuilder
+    extends SpringCredentialBuilderBase<PrefixedSpringEnvironmentCredentialBuilder> {
```
Comment (chenrujun) [inline sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/PrefixedSpringEnvironmentCredentialBuilder.java:13]: Same here: I suggest to delete this builder, use constructor instead.

File: sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/SpringEnvironmentCredential.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+            return Mono.error(() -> {
+                                  final CredentialUnavailableException exception = new CredentialUnavailableException(
+                                      "SpringEnvironmentCredential authentication unavailable. Environment variables "
+                                          + "are not fully configured.");
+                                  LOGGER.error(exception.getMessage(), exception);
+                                  return exception;
+                              }
+            );
+        } else {
+            return tokenCredential.getToken(request);
```
Comment (chenrujun) [inline sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/SpringEnvironmentCredential.java:63]: Same here: Duplicated codes.

File: sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/SpringEnvrionmentCredentialBuilder.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.spring.identity;
+
+import org.springframework.core.env.Environment;
+
+public class SpringEnvrionmentCredentialBuilder
```
Comment (chenrujun) [inline sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/SpringEnvrionmentCredentialBuilder.java:8]: typo: `Envrionment` -> `Environment`.

Comment (chenrujun) [reply sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/SpringEnvrionmentCredentialBuilder.java:8]: And this class is never used.

File: sdk/spring/azure-spring-cloud-context/src/main/java/com/azure/spring/cloud/autoconfigure/context/AzureContextAutoConfiguration.java [hunk 10]
```diff
@@ -39,23 +44,25 @@
...
     }
 
     @Bean
     @ConditionalOnMissingBean
     public TokenCredential credential(Environment environment) {
-        return new DefaultSpringCredentialBuilder().environment(environment)
-                                                   .alternativePrefix(AzureProperties.PREFIX)
-                                                   .build();
+        final List<TokenCredential> tokenCredentials = ((DefaultSpringCredential) new DefaultSpringCredentialBuilder(
+            environment).build()).getTokenCredentials();
```
Comment (chenrujun) [inline sdk/spring/azure-spring-cloud-context/src/main/java/com/azure/spring/cloud/autoconfigure/context/AzureContextAutoConfiguration.java:61]: If we use constructor, code will be like this:
```
new DefaultSpringCredential(environment).getTokenCredentials();
```
Much easier to read.

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/AzureEnvironment.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    public String getClientId() {
+        return getPropertyValue(AZURE_CREDENTIAL_PREFIX + CLIENT_ID,
+                                sdkEnvValue(Configuration.PROPERTY_AZURE_CLIENT_ID));
+    }
+
+    public String getClientSecret() {
+        return getPropertyValue(AZURE_CREDENTIAL_PREFIX + CLIENT_SECRET,
+                                sdkEnvValue(Configuration.PROPERTY_AZURE_CLIENT_ID));
```
[the remark this answers is not in the document]
Comment (stliu) [reply sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/AzureEnvironment.java:39]: nice catch

File: sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/DefaultSpringCredentialBuilder.java (hunk 1 above)
Comment (saragluna) [reply sdk/spring/azure-identity-spring/src/main/java/com/azure/spring/identity/DefaultSpringCredentialBuilder.java]: just followed the sdk builder pattern.

**Comment by saragluna** [comment]
Close in favor of #22938.

**Pull request closed without merging**